lib/datadog/grok/src/parse_grok.rs

```diff
@@ -0,0 +1,289 @@
+use std::collections::HashMap;
+
+use lazy_static::lazy_static;
+use regex::Regex;
+
+use lookup::LookupBuf;
+
+use crate::ast;
+use crate::ast::{Function, GrokPattern};
+use crate::parse_grok_pattern::parse_grok_pattern;
+use vrl::Value;
+
+#[derive(Debug, Clone)]
+pub struct GrokRule {
+    pub pattern: String,
+    pub filters: HashMap<LookupBuf, Vec<ast::Function>>,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Failed to parse grok expression: '{}'", .0)]
+    InvalidGrokExpression(String),
+    #[error("Invalid arguments for the function '{}'", .0)]
+    InvalidFunctionArguments(String),
+    #[error("Unsupported filter '{}'", .0)]
+    UnsupportedFilter(String),
+}
+
+/**
+Parses DD grok rules.
+
+Here is an example:
+access.common %{_client_ip} %{_ident} %{_auth} \[%{_date_access}\] "(?>%{_method} |)%{_url}(?> %{_version}|)" %{_status_code} (?>%{_bytes_written}|-)
+access.combined %{access.common} (%{number:duration:scale(1000000000)} )?"%{_referer}" "%{_user_agent}"( "%{_x_forwarded_for}")?.*"#
+
+You can write parsing rules with the %{MATCHER:EXTRACT:FILTER} syntax:
+- Matcher: A rule (possibly a reference to another token rule) that describes what to expect (number, word, notSpace, etc.)
+- Extract (optional): An identifier representing the capture destination for the piece of text matched by the Matcher.
+- Filter (optional): A post-processor of the match to transform it.
+
+Each rule can reference parsing rules defined above itself in the list.
+Only one can match any given log. The first one that matches, from top to bottom, is the one that does the parsing.
+For further documentation and the full list of available matcher and filters check out https://docs.datadoghq.com/logs/processing/parsing
+*/
+pub fn parse_grok_rules(
+    support_rules: &Vec<String>,
+    match_rules: &Vec<String>,
+) -> Result<Vec<GrokRule>, Error> {
+    let mut prev_rule_definitions: HashMap<&str, String> = HashMap::new();
+    let mut prev_rule_destinations: HashMap<&str, HashMap<LookupBuf, Vec<ast::Function>>> =
+        HashMap::new();
+
+    // parse support rules to reference them later in the match rules
+    support_rules
+        .iter()
+        .filter(|&r| !r.is_empty())
+        .map(|r| parse_grok_rule(r, &mut prev_rule_definitions, &mut prev_rule_destinations))
+        .collect::<Result<Vec<GrokRule>, Error>>()?;
```

Comment:
I'm not clear on the purpose of this block. It appears to be collecting to a vector of `GrokRule`, which is then discarded?

If the purpose is to exit early on an invalid support rule, I'd favour removing the allocation and continuation of subsequent errors, with something like:

```rust
for rule in support_rules.iter().filter(|&r| !r.is_empty()) {
    parse_grok_rule(
        rule,
        &mut prev_rule_definitions,
        &mut prev_rule_destinations,
    )?;
}
```